phpBB/phpbb/install/controller/helper.php

```diff
@@ -0,0 +1,363 @@
+	 * @param template				$template
+	 * @param path_helper			$path_helper
+	 * @param request				$phpbb_request
+	 * @param symfony_request		$request
+	 * @param router				$router
+	 * @param string				$phpbb_root_path
+	 */
+	public function __construct(language $language, language_file_helper $lang_helper, navigation_provider $nav, template $template, path_helper $path_helper, request $phpbb_request, symfony_request $request, router $router, $phpbb_root_path)
+	{
+		$this->language = $language;
+		$this->language_cookie = false;
+		$this->lang_helper = $lang_helper;
+		$this->navigation_provider = $nav;
+		$this->template = $template;
+		$this->path_helper = $path_helper;
+		$this->phpbb_request = $phpbb_request;
+		$this->request = $request;
+		$this->router = $router;
+		$this->phpbb_root_path = $phpbb_root_path;
+		$this->phpbb_admin_path = $phpbb_root_path . 'adm/';
+
+		$this->handle_language_select();
+	}
+
+	/**
+	 * Automate setting up the page and creating the response object.
+	 *
+	 * @param string	$template_file		The template handle to render
+	 * @param string	$page_title			The title of the page to output
+	 * @param bool		$selected_language	True to enable language selector it, false otherwise
+	 * @param int		$status_code		The status code to be sent to the page header
+	 *
+	 * @return Response object containing rendered page
+	 */
+	public function render($template_file, $page_title = '', $selected_language = false, $status_code = 200)
+	{
+		$this->page_header($page_title, $selected_language);
+
+		$this->template->set_filenames(array(
+			'body'	=> $template_file,
+		));
+
+		$response = new Response($this->template->assign_display('body'), $status_code);
+
+		// Set language cookie
+		if ($this->language_cookie !== false)
+		{
+			$cookie = new Cookie('lang', $this->language_cookie, time() + 3600);
+			$response->headers->setCookie($cookie);
+
+			$this->language_cookie = false;
+		}
+
+		return $response;
+	}
+
+	/**
+	 * Returns path from route name
+	 *
+	 * @param string	$route_name
+	 *
+	 * @return string
+	 */
+	public function route($route_name)
+	{
+		$url = $this->router->generate($route_name);
+
+		return $url;
+	}
+
+	/**
+	 * Handles language selector form
+	 */
+	protected function handle_language_select()
+	{
+		$lang = null;
+
+		// Check if language form has been submited
+		$submit = $this->phpbb_request->variable('change_lang', '');
+		if (!empty($submit))
+		{
+			$lang = $this->phpbb_request->variable('language', '');
+
+			if (!empty($lang))
+			{
+				$this->language_cookie = $lang;
+			}
+		}
+
+		// Retrive language from cookie
+		$lang_cookie = $this->phpbb_request->variable('lang', '', false, request_interface::COOKIE);
+		if (empty($lang) && !empty($lang_cookie))
+		{
+			$lang = $lang_cookie;
+			$this->language_cookie = $lang;
+		}
+
+		$lang = (!empty($lang)) ? $lang : null;
+		$this->render_language_select($lang);
+
+		if ($lang !== null)
+		{
+			$this->language->set_user_language($lang, true);
```

Comment:
This might be a vulnerability, it should be checked at least if the result does not contain any path changing characters. Even better that the directory is valid etc.